File: src/resilience/bulkhead.rs
//! Bulkhead pattern implementation for Hivemind
//!
//! This module provides an implementation of the Bulkhead pattern,
//! which limits the number of concurrent requests to a service to prevent
//! cascading failures.

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::sync::{Mutex, Semaphore};
use tracing::{debug, error, info, warn};

/// Bulkhead configuration
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct BulkheadConfig {
    /// Maximum number of concurrent requests
    pub max_concurrent_calls: u32,
    /// Maximum number of queued requests
    pub max_queue_size: u32,
    /// Timeout for acquiring a permit (milliseconds)
    pub acquire_timeout_ms: u64,
}

impl Default for BulkheadConfig {
    fn default() -> Self {
        Self {
            max_concurrent_calls: 10,
            max_queue_size: 20,
            acquire_timeout_ms: 1000,
        }
    }
}

/// Bulkhead metrics
#[derive(Clone, Debug, Serialize, Deserialize, Default)]
pub struct BulkheadMetrics {
    /// Total number of successful requests
    pub successful_calls: u64,
    /// Total number of rejected requests
    pub rejected_calls: u64,
    /// Total number of timed out requests
    pub timed_out_calls: u64,
    /// Current number of active requests
    pub active_calls: u32,
    /// Current number of queued requests
    pub queued_calls: u32,
    /// Maximum number of concurrent calls observed
    pub max_concurrent_calls_observed: u32,
    /// Maximum number of queued calls observed
    pub max_queued_calls_observed: u32,
    /// Average wait time (milliseconds)
    pub avg_wait_time_ms: f64,
    /// Average execution time (milliseconds)
    pub avg_execution_time_ms: f64,
}

/// Bulkhead implementation
#[derive(Clone)]
pub struct Bulkhead {
    /// Service name
    service_name: String,
    /// Bulkhead configuration
    config: Arc<Mutex<BulkheadConfig>>,
    /// Bulkhead metrics
    metrics: Arc<Mutex<BulkheadMetrics>>,
    /// Semaphore for limiting concurrent requests
    semaphore: Arc<Semaphore>,
}

impl Bulkhead {
    /// Create a new bulkhead
    pub fn new(service_name: String, config: BulkheadConfig) -> Self {
        let semaphore = Arc::new(Semaphore::new(config.max_concurrent_calls as usize));
        
        Self {
            service_name,
            config: Arc::new(Mutex::new(config)),
            metrics: Arc::new(Mutex::new(BulkheadMetrics::default())),
            semaphore,
        }
    }
    
    /// Acquire permission to execute a request
    pub async fn acquire_permission(&self) -> bool {
        let start_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        
        let config = self.config.lock().await.clone();
        
        // Try to acquire a permit with timeout
        let result = tokio::time::timeout(
            Duration::from_millis(config.acquire_timeout_ms),
            self.semaphore.acquire(),
        ).await;
        
        let end_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        
        let wait_time = end_time - start_time;
        
        let mut metrics = self.metrics.lock().await;
        
        match result {
            Ok(Ok(permit)) => {
                // Successfully acquired a permit
                permit.forget(); // Don't release the permit automatically
                
                metrics.active_calls += 1;
                metrics.queued_calls = metrics.queued_calls.saturating_sub(1);
                
                // Update max observed values
                if metrics.active_calls > metrics.max_concurrent_calls_observed {
                    metrics.max_concurrent_calls_observed = metrics.active_calls;
                }
                
                // Update average wait time
                let total_calls = metrics.successful_calls + metrics.rejected_calls + metrics.timed_out_calls;
                if total_calls > 0 {
                    metrics.avg_wait_time_ms = ((metrics.avg_wait_time_ms * (total_calls as f64)) + (wait_time as f64)) / ((total_calls + 1) as f64);
                } else {
                    metrics.avg_wait_time_ms = wait_time as f64;
                }
                
                debug!(
                    "Bulkhead for service {} acquired permission (active: {}, max: {})",
                    self.service_name, metrics.active_calls, config.max_concurrent_calls
                );
                
                true
            },
            Ok(Err(_)) => {
                // Failed to acquire a permit (semaphore closed)
                metrics.rejected_calls += 1;
                
                warn!(
                    "Bulkhead for service {} rejected request (semaphore closed)",
                    self.service_name
                );
                
                false
            },
            Err(_) => {
                // Timeout waiting for a permit
                metrics.timed_out_calls += 1;
                
                warn!(
                    "Bulkhead for service {} timed out waiting for permission after {}ms",
                    self.service_name, config.acquire_timeout_ms
                );
                
                false
            }
        }
    }
    
    /// Release permission after request completion
    pub async fn release_permission(&self) {
        let mut metrics = self.metrics.lock().await;
        
        // Add a permit back to the semaphore
        self.semaphore.add_permits(1);
        
        // Update metrics
        metrics.active_calls = metrics.active_calls.saturating_sub(1);
        metrics.successful_calls += 1;
        
        debug!(
            "Bulkhead for service {} released permission (active: {})",
            self.service_name, metrics.active_calls
        );
    }
    
    /// Get current metrics
    pub async fn get_metrics(&self) -> BulkheadMetrics {
        self.metrics.lock().await.clone()
    }
    
    /// Update configuration
    pub async fn update_config(&self, config: BulkheadConfig) {
        let mut current_config = self.config.lock().await;
        
        // Update the semaphore if max_concurrent_calls changed
        if config.max_concurrent_calls != current_config.max_concurrent_calls {
            let current_permits = self.semaphore.available_permits();
            let diff = config.max_concurrent_calls as isize - current_config.max_concurrent_calls as isize;
            
            if diff > 0 {
                // Add permits
                self.semaphore.add_permits(diff as usize);
            } else if diff < 0 {
                // Remove permits (up to available)
                let to_remove = (-diff as usize).min(current_permits);
                for _ in 0..to_remove {
                    if let Ok(permit) = self.semaphore.try_acquire() {
                        permit.forget();
                    }
                }
            }
        }
        
        // Update config
        *current_config = config;
        
        info!(
            "Bulkhead for service {} updated configuration (max_concurrent_calls: {}, max_queue_size: {}, acquire_timeout_ms: {})",
            self.service_name, 
            current_config.max_concurrent_calls,
            current_config.max_queue_size,
            current_config.acquire_timeout_ms
        );
    }
    
    /// Reset the bulkhead
    pub async fn reset(&self) {
        // Reset metrics
        *self.metrics.lock().await = BulkheadMetrics::default();
        
        // Reset semaphore
        let config = self.config.lock().await;
        let current_permits = self.semaphore.available_permits();
        let max_permits = config.max_concurrent_calls as usize;
        
        if current_permits < max_permits {
            // Add missing permits
            self.semaphore.add_permits(max_permits - current_permits);
        } else if current_permits > max_permits {
            // Remove excess permits
            let to_remove = current_permits - max_permits;
            for _ in 0..to_remove {
                if let Ok(permit) = self.semaphore.try_acquire() {
                    permit.forget();
                }
            }
        }
        
        info!("Bulkhead for service {} reset", self.service_name);
    }
    
    /// Execute a function with bulkhead protection
    pub async fn execute<F, Fut, T, E>(&self, f: F) -> Result<T, E>
    where
        F: FnOnce() -> Fut,
        Fut: std::future::Future<Output = Result<T, E>>,
        E: std::error::Error,
    {
        // Try to acquire permission
        if !self.acquire_permission().await {
            return Err(E::from_error("Bulkhead rejected request"));
        }
        
        let start_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        
        // Execute the function
        let result = f().await;
        
        let end_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_millis() as u64;
        
        let execution_time = end_time - start_time;
        
        // Update metrics
        let mut metrics = self.metrics.lock().await;
        let total_calls = metrics.successful_calls + metrics.rejected_calls + metrics.timed_out_calls;
        if total_calls > 0 {
            metrics.avg_execution_time_ms = ((metrics.avg_execution_time_ms * (total_calls as f64)) + (execution_time as f64)) / ((total_calls + 1) as f64);
        } else {
            metrics.avg_execution_time_ms = execution_time as f64;
        }
        drop(metrics);
        
        // Release permission
        self.release_permission().await;
        
        result
    }
}

/// Extension trait for error types
impl<E: std::error::Error + 'static> crate::resilience::RetryableError for E {
    fn from_error(msg: &str) -> Self {
        std::io::Error::new(std::io::ErrorKind::Other, msg).into()
    }
}

File: internal/resilience/bulkhead.go
// Package resilience provides resilience patterns for the Hivemind system.
package resilience

import (
	"context"
	"sync"
	"sync/atomic"
	"time"

	"go.uber.org/zap"
	"golang.org/x/sync/semaphore"
)

// BulkheadConfig defines the configuration for a bulkhead
type BulkheadConfig struct {
	// MaxConcurrentCalls is the maximum number of concurrent requests
	MaxConcurrentCalls int64
	// MaxQueueSize is the maximum number of queued requests
	MaxQueueSize int64
	// AcquireTimeoutMs is the timeout for acquiring a permit (milliseconds)
	AcquireTimeoutMs int64
}

// DefaultBulkheadConfig returns the default bulkhead configuration
func DefaultBulkheadConfig() *BulkheadConfig {
	return &BulkheadConfig{
		MaxConcurrentCalls: 10,
		MaxQueueSize:       20,
		AcquireTimeoutMs:   1000,
	}
}

// BulkheadMetrics contains metrics for a bulkhead
type BulkheadMetrics struct {
	// SuccessfulCalls is the total number of successful requests
	SuccessfulCalls uint64
	// RejectedCalls is the total number of rejected requests
	RejectedCalls uint64
	// TimedOutCalls is the total number of timed out requests
	TimedOutCalls uint64
	// ActiveCalls is the current number of active requests
	ActiveCalls int64
	// QueuedCalls is the current number of queued requests
	QueuedCalls int64
	// MaxConcurrentCallsObserved is the maximum number of concurrent calls observed
	MaxConcurrentCallsObserved int64
	// MaxQueuedCallsObserved is the maximum number of queued calls observed
	MaxQueuedCallsObserved int64
	// AvgWaitTimeMs is the average wait time (milliseconds)
	AvgWaitTimeMs float64
	// AvgExecutionTimeMs is the average execution time (milliseconds)
	AvgExecutionTimeMs float64
}

// Bulkhead implements the bulkhead pattern to limit concurrent requests
type Bulkhead struct {
	// ServiceName is the name of the service
	serviceName string
	// Config is the bulkhead configuration
	config atomic.Pointer[BulkheadConfig]
	// Metrics is the bulkhead metrics
	metrics *BulkheadMetrics
	// Semaphore for limiting concurrent requests
	semaphore *semaphore.Weighted
	// Mutex for protecting metrics updates
	mu sync.Mutex
	// Logger
	logger *zap.Logger
}

// NewBulkhead creates a new bulkhead
func NewBulkhead(serviceName string, config BulkheadConfig) *Bulkhead {
	logger, _ := zap.NewProduction()

	b := &Bulkhead{
		serviceName: serviceName,
		metrics:     &BulkheadMetrics{},
		semaphore:   semaphore.NewWeighted(config.MaxConcurrentCalls),
		logger:      logger,
	}

	// Store config in atomic pointer
	b.config.Store(&config)

	return b
}

// AcquirePermission acquires permission to execute a request
func (b *Bulkhead) AcquirePermission(ctx context.Context) bool {
	startTime := time.Now()

	config := b.config.Load()

	// Create a context with timeout
	acquireCtx, cancel := context.WithTimeout(ctx, time.Duration(config.AcquireTimeoutMs)*time.Millisecond)
	defer cancel()

	// Try to acquire a permit with timeout
	err := b.semaphore.Acquire(acquireCtx, 1)

	waitTime := time.Since(startTime).Milliseconds()

	b.mu.Lock()
	defer b.mu.Unlock()

	if err == nil {
		// Successfully acquired a permit
		b.metrics.ActiveCalls++
		if b.metrics.QueuedCalls > 0 {
			b.metrics.QueuedCalls--
		}

		// Update max observed values
		if b.metrics.ActiveCalls > b.metrics.MaxConcurrentCallsObserved {
			b.metrics.MaxConcurrentCallsObserved = b.metrics.ActiveCalls
		}

		// Update average wait time
		totalCalls := b.metrics.SuccessfulCalls + b.metrics.RejectedCalls + b.metrics.TimedOutCalls
		if totalCalls > 0 {
			b.metrics.AvgWaitTimeMs = ((b.metrics.AvgWaitTimeMs * float64(totalCalls)) + float64(waitTime)) / float64(totalCalls+1)
		} else {
			b.metrics.AvgWaitTimeMs = float64(waitTime)
		}

		b.logger.Debug("Bulkhead acquired permission",
			zap.String("service", b.serviceName),
			zap.Int64("active", b.metrics.ActiveCalls),
			zap.Int64("max", config.MaxConcurrentCalls))

		return true
	} else if err == context.DeadlineExceeded {
		// Timeout waiting for a permit
		b.metrics.TimedOutCalls++

		b.logger.Warn("Bulkhead timed out waiting for permission",
			zap.String("service", b.serviceName),
			zap.Int64("timeout_ms", config.AcquireTimeoutMs))

		return false
	} else {
		// Failed to acquire a permit (context canceled or other error)
		b.metrics.RejectedCalls++

		b.logger.Warn("Bulkhead rejected request",
			zap.String("service", b.serviceName),
			zap.Error(err))

		return false
	}
}

// ReleasePermission releases permission after request completion
func (b *Bulkhead) ReleasePermission() {
	b.mu.Lock()
	defer b.mu.Unlock()

	// Release a permit
	b.semaphore.Release(1)

	// Update metrics
	if b.metrics.ActiveCalls > 0 {
		b.metrics.ActiveCalls--
	}
	b.metrics.SuccessfulCalls++

	b.logger.Debug("Bulkhead released permission",
		zap.String("service", b.serviceName),
		zap.Int64("active", b.metrics.ActiveCalls))
}

// GetMetrics returns the current metrics
func (b *Bulkhead) GetMetrics() BulkheadMetrics {
	b.mu.Lock()
	defer b.mu.Unlock()

	return *b.metrics
}

// UpdateConfig updates the bulkhead configuration
func (b *Bulkhead) UpdateConfig(config BulkheadConfig) {
	oldConfig := b.config.Load()

	// Update the semaphore if max_concurrent_calls changed
	if config.MaxConcurrentCalls != oldConfig.MaxConcurrentCalls {
		diff := config.MaxConcurrentCalls - oldConfig.MaxConcurrentCalls

		if diff > 0 {
			// Add permits
			b.semaphore.Release(diff)
		} else if diff < 0 {
			// Try to remove permits (up to available)
			ctx := context.Background()
			if err := b.semaphore.Acquire(ctx, -diff); err == nil {
				// Successfully acquired permits to remove
				// We don't release them, effectively reducing the semaphore size
			}
		}
	}

	// Update config
	b.config.Store(&config)

	b.logger.Info("Bulkhead updated configuration",
		zap.String("service", b.serviceName),
		zap.Int64("max_concurrent_calls", config.MaxConcurrentCalls),
		zap.Int64("max_queue_size", config.MaxQueueSize),
		zap.Int64("acquire_timeout_ms", config.AcquireTimeoutMs))
}

// Reset resets the bulkhead
func (b *Bulkhead) Reset() {
	b.mu.Lock()
	defer b.mu.Unlock()

	// Reset metrics
	b.metrics = &BulkheadMetrics{}

	// Reset semaphore
	config := b.config.Load()

	// Create a new semaphore with the configured size
	b.semaphore = semaphore.NewWeighted(config.MaxConcurrentCalls)

	b.logger.Info("Bulkhead reset", zap.String("service", b.serviceName))
}

// Execute executes a function with bulkhead protection
func (b *Bulkhead) Execute(ctx context.Context, f func(ctx context.Context) (interface{}, error)) (interface{}, error) {
	var zero interface{}

	// Try to acquire permission
	if !b.AcquirePermission(ctx) {
		return zero, ErrBulkheadFull
	}

	startTime := time.Now()

	// Execute the function
	result, err := f(ctx)

	executionTime := time.Since(startTime).Milliseconds()

	// Update metrics
	b.mu.Lock()
	totalCalls := b.metrics.SuccessfulCalls + b.metrics.RejectedCalls + b.metrics.TimedOutCalls
	if totalCalls > 0 {
		b.metrics.AvgExecutionTimeMs = ((b.metrics.AvgExecutionTimeMs * float64(totalCalls)) + float64(executionTime)) / float64(totalCalls+1)
	} else {
		b.metrics.AvgExecutionTimeMs = float64(executionTime)
	}
	b.mu.Unlock()

	// Release permission
	b.ReleasePermission()

	return result, err
}
